### PR TITLE
Feat/370 improve price sync since beginning

### PR DESCRIPTION
**Pull request by vorg-code**
Test Plan:
1. To test building all prices from scratch, run (warning, this will reset your entire local database):  yarn docker mr -f
      + you should see a new file created: prisma/seeds/prices.csv, it has all prices since the beginning until today (this file will be created only once)
      + maybe you see some lines “Problem getting price of XEC 20210606 -> Error: timeout of 30000ms exceeded (attempt 2)”, this is only a problem if you see attempt 5, any lesser will be retried.
      + you can query your local database and see all thousands of prices there: select * from paybutton.Price
2. To test building prices from your file you can run again:  yarn docker mr -f
      + It will run much faster and you won’t see any attempt to reach coin dance api
3. To test the price worker sync, you can open the prices.csv, delete the 3 last XEC lines (each line is a day) and:
    1. Run again: yarn docker mr -f, this will recreate your database, 3 last days will be missing, you can verify that querying the database:
        select id, FROM_UNIXTIME(p.`timestamp`), p.value, p.quoteId, p.networkId 
        from paybutton.Price p 
        where networkId = 1
        order by p.`timestamp` desc
    2. Run the worker, 1st terminal: yarn docker jr (run the worker), 2nd terminal: yarn docker jw (see the worker output)
    3. Run query in 1 again, you will see the prices there

With both the big first pull and the worker that runs every some seconds, we will always have all the price data ready in our database

### CODE REVIEW

**Review by Klakurka — CHANGES_REQUESTED** [review]
Doesn't build for me:

image

**Comment by vorg-code** [comment]
ah I forgot to mention that you have to set the PRICE_API_URL in a .env.development.local file (PRICE_API_URL=https://...)

**Review by ScottMcDermid — CHANGES_REQUESTED** [review]
Currently, running the script sends close to 3000 http requests within seconds. Can we throttle the request rate so that we handle a few requests at a time?

**Comment by Klakurka** [comment]
Even better than sending 3000 requests would be just to send 1 and have it to return what we need.

What would be best there? We have total control over the CD API. I just want to make the endpoint relatively generic still if possible.

**Comment by Klakurka** [comment]
> Currently, running the script sends close to 3000 http requests within seconds. Can we throttle the request rate so that we handle a few requests at a time?

That's what this task refers to: https://github.com/PayButton/paybutton-server/issues/403

I'll leave it to you guys to decide if we should try to improve that after merging this or as part of this MR.

**Comment by vorg-code** [comment]
@Klakurka I created #403 after creating this PR, the intention was for this PR to be merged and then I would create a new PR to improve the transactions' performance, since transactions and prices can be seen as different modules. 

What? CD is ours? One endpoint call for each coin to get the whole price history would be the simplest and fastest and still quite useful generically speaking imo. Should I have access to that github repo, I could implement it as well.

**Comment by Klakurka** [comment]
I can add it.

File: prisma/seeds/prices.ts [hunk 1]
```diff
@@ -1,35 +1,145 @@
...
+      for (let j = 0; j < csvContent[0].length; j++) { // headers
+        newPrice[csvContent[0][j]] = csvContent[i][j]
+      }
+      // ignore some prices that come empty
+      if (newPrice.priceInCAD !== '' && newPrice.priceInUSD !== '') { res.push(newPrice as PriceFileData) }
+    }
+
+    return res
+  } else {
+    if (attempt > 3) { throw new Error('Already attempted 3 times to create price files without success') }
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply prisma/seeds/prices.ts:119]: done

File: services/priceService.ts [hunk 2]
```diff
@@ -62,23 +62,76 @@ export async function upsertCurrentPricesForNetworkId (responseData: IResponseDa
...
+  try {
+    const res = await axios.get(`${appInfo.priceAPIURL}/${ticker}+${dayString}`, {
+      timeout: 30 * 1000
+    })
+
+    if (res.data.success !== false) { return res.data } else { return null }
+  } catch (error) {
+    console.error(`Problem getting price of ${ticker} ${day.format('YYYYMMDD')} -> ${error as string} (attempt ${attempt})`)
+
+    if (attempt < 5) { return await getPriceForDayTicker(day, ticker, ++attempt) } else { return null }
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply services/priceService.ts:87]: indeed much better, done

File: prisma/seeds/prices.ts [hunk 3]
```diff
@@ -1,35 +1,145 @@
...
+
+    await createPricesFile()
+    return await getPricesFromFile(++attempt)
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply prisma/seeds/prices.ts:122]: done

**Comment by vorg-code** [comment]
Changes addressed. Using this new endpoint is much faster, 2 seconds only.

Now we have to change .env.development.local to have both these (more flexibility for different api endpoints): 
- PRICE_API_URL=[api url until 'api/']
- PRICE_API_TOKEN=[only the token]

**Comment by Klakurka** [comment]
> Changes addressed. Using this new endpoint is much faster, 2 seconds only.
> 
> Now we have to change .env.development.local to have both these (more flexibility for different api endpoints):
> 
> * PRICE_API_URL=[api url until 'api/']
> * PRICE_API_TOKEN=[only the token]

It doesn't matter right now but we do kinda want this to be straight forward forward for others to fork the project and hook up to a different API. I guess we can v2 that but right now it's looking pretty tailored to the CD API.

**Review by Klakurka — DISMISSED** [review]

File: services/priceService.ts [hunk 4]
```diff
@@ -66,25 +65,70 @@ export async function upsertCurrentPricesForNetworkId (responseData: IResponseDa
...
+}
+
+export async function getAllPricesByTicker (ticker: string, attempt: number = 1): Promise<IResponseDataDaily[] | null> {
+  try {
+    const res = await axios.get(getAllPricesByTickerURL(ticker), {
+      timeout: PRICE_API_TIMEOUT
+    })
+
+    if (res.data.success !== false) {
+      const dailyPrices: IResponseDataDaily[] = Object.entries<IResponseData>(res.data).map(e => {
```
Comment (ScottMcDermid) [inline services/priceService.ts:118]: Destructure the `e` argument to have more descriptive values than `e[0]` and `e[1]`.
Something like 
```js
.map( ([day, priceData]) => ({ day, ...priceData }) )
```

File: services/priceService.ts [hunk 5]
```diff
@@ -66,25 +65,70 @@ export async function upsertCurrentPricesForNetworkId (responseData: IResponseDa
...
     throw new Error(RESPONSE_MESSAGES.MISSING_PRICE_API_URL_400.message)
   }
-  if (!Object.values(TICKERS).includes(ticker)) {
-    throw new Error(RESPONSE_MESSAGES.INVALID_TICKER_400.message)
+  if (appInfo.priceAPIToken === '') {
+    throw new Error(RESPONSE_MESSAGES.MISSING_PRICE_API_TOKEN_400.message)
   }
+}
+
+function getPriceByDayURL (day: moment.Moment, ticker: string): string {
```
Comment (ScottMcDermid) [inline services/priceService.ts:83]: This function sounds like it returns a price value. Maybe reword to something like `getURLForPriceByDayRequest` or something like that

File: services/priceService.ts [hunk 6]
```diff
@@ -62,23 +61,121 @@ export async function upsertCurrentPricesForNetworkId (responseData: IResponseDa
...
+    if (attempt < PRICE_API_MAX_RETRIES) {
+      return await getAllPricesByTicker(ticker, attempt + 1)
+    } else { return null }
+  }
+}
+
+export async function syncPastPrices (): Promise<void> {
+  const lastTimestamp = await prisma.price.findFirst({
+    where: {
+      networkId: XEC_NETWORK_ID, // any valid network
```
Comment (ScottMcDermid) [inline services/priceService.ts:138]: If we don't care about `networkId` and `quoteId`, can we just omit the `where` clause entirely?

**Review by ScottMcDermid — CHANGES_REQUESTED** [review]
Building the price file for the first time works for me, and it loads into the db as expected.
Rebuilding the price file works for me (step 2)
Attempting to rebuild my price file after deleting the last 3 lines of `prices.csv` gives me an error (step 3 of the test plan):

```
✔ Generated Prisma Client (3.14.0 | library) to ./node_modules/@prisma/client in 225ms

Running seed command `ts-node --compiler-options {"module":"CommonJS"} -r tsconfig-paths/register prisma/seed.ts` ...
Error: [DecimalError] Invalid argument: undefined
    at new Decimal (/home/node/src/node_modules/@prisma/client/runtime/index.js:38450:13)
    at /home/node/src/prisma/seeds/prices.ts:116:16
    at Array.map (<anonymous>)
    at /home/node/src/prisma/seeds/prices.ts:113:12
    at Array.map (<anonymous>)
    at getPrices (/home/node/src/prisma/seeds/prices.ts:112:44)
    at main (/home/node/src/prisma/seed.ts:46:43)

An error occured while running the seed command:
Error: Command failed with exit code 1: ts-node --compiler-options {"module":"CommonJS"} -r tsconfig-paths/register prisma/seed.ts
```

And this appears in my job log:

```
grpc-bchrpc-node: failed network integrity check
grpc-bchrpc-node: failed network integrity check
grpc-bchrpc-node: failed network integrity check
automatic syncing of address FAILED
error for automatic syncing of addresses: 
Invalid `prisma.transaction.upsert()` invocation:


  Foreign key constraint failed on the field: `addressId`
job repeat:6770ef6bb5c4384db03b15ad5b469ac8:1676585880000: syncing current prices...
/home/node/src/node_modules/@prisma/client/runtime/index.js:45405
        throw new PrismaClientKnownRequestError(message, e.code, this.client._clientVersion, e.meta);
              ^


PrismaClientKnownRequestError: 
Invalid `prisma.price.upsert()` invocation:


  Foreign key constraint failed on the field: `networkId`
    at Object.request (/home/node/src/node_modules/@prisma/client/runtime/index.js:45405:15)
    at PrismaClient._request (/home/node/src/node_modules/@prisma/client/runtime/index.js:46301:18)
    at upsertPricesForNetworkId (/home/node/src/services/priceService.ts:26:3)
    at upsertCurrentPricesForNetworkId (/home/node/src/services/priceService.ts:65:3) {
  code: 'P2003',
  clientVersion: '3.14.0',
  meta: { field_name: 'networkId' }
}
```

**Review by chedieck — COMMENTED** [review]
Left some comments, but I have some other points:
1. The PR is considerably big, which will difficult the review process. Since we do [trunk-based development](https://web.archive.org/web/20220512072646/https://ourmachinery.com/post/step-by-step-programming-incrementally/), is better to always divide the PRs as better as we can and avoid doing PRs that are completely features.
2. I also feel like that a lot of the logic about retrying the sync in case some of it fails should be done in the `jobs/workers.ts` file. There is logic in bullmq to handle repeating jobs that fail (or not), and I think it would be easier to understand what is happenning if the logic was something like:

----

1. Check the CSV file for missing prices.
2. If they are all there, do nothing and don't run the job until tomorrow
4. If something is missing, fetch them, try to put them on the DB and repeat step 1 in X seconds.
----
The current implementations includes multiple uses of `attempts` that go into functions that are recursive simply because they can fail. For that reason, I had to use a lot of brain power and code-jumping to understand the logic flow, which (as far as I understood), include multiple 'concurrent attempts', some that result in an error upon failure, some that are ignored. 

Finally, except on `grpcServices` that always use 3rd party services by definition, is better if we mantain the convention of prefixing with `sync` functions that access 3rd party services.


Finally,

File: services/priceService.ts [hunk 7]
```diff
@@ -62,23 +62,76 @@ export async function upsertCurrentPricesForNetworkId (responseData: IResponseDa
...
   }
-  const todayString = moment().format(PRICE_API_DATE_FORMAT)
+  if (!Object.values(TICKERS).includes(ticker)) {
+    throw new Error(RESPONSE_MESSAGES.INVALID_TICKER_400.message)
+  }
+
+  const dayString = day.format(PRICE_API_DATE_FORMAT)
+  try {
+    const res = await axios.get(`${appInfo.priceAPIURL}/${ticker}+${dayString}`, {
+      timeout: 30 * 1000
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply services/priceService.ts]: done

File: services/priceService.ts (hunk 4 above)
Comment (vorg-code) [reply services/priceService.ts:118]: nice, done

File: services/priceService.ts (hunk 6 above)
Comment (vorg-code) [reply services/priceService.ts:138]: indeed! done

File: constants/index.ts [hunk 8]
```diff
@@ -30,16 +30,25 @@ export const RESPONSE_MESSAGES = {
...
   NETWORK_SLUG_NOT_PROVIDED_400: { statusCode: 400, message: "'networkSlug' not provided." },
   QUOTE_SLUG_NOT_PROVIDED_400: { statusCode: 400, message: "'quoteSlug' not provided." },
   NO_CURRENT_PRICES_FOUND_404: { statusCode: 404, message: 'Current prices not found.' },
-  INVALID_QUOTE_SLUG_400: { statusCode: 400, message: 'Invalid quote slug.' }
+  INVALID_QUOTE_SLUG_400: { statusCode: 400, message: 'Invalid quote slug.' },
+  INVALID_TICKER_400: { statusCode: 400, message: 'Invalid ticker.' }
 }
 
-export const NETWORK_SLUGS = {
+export interface KeyValueAny {
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply constants/index.ts:44]: much more clean and generic solution, thanks and done!

File: constants/index.ts [hunk 9]
```diff
@@ -72,10 +81,21 @@ export const SUPPORTED_QUOTES = [ // avoids hitting the DB every time for data t
...
 // Wait time (in ms) between sync of current prices
 export const CURRENT_PRICE_SYNC_DELAY = 60000
 export const DEFAULT_WORKER_LOCK_DURATION = 120000
+
+export const TICKERS: KeyValueString = {
+  ecash: 'XEC',
+  bitcoincash: 'BCH'
+}
+
+export const NETWORKS: KeyValueAny = { XEC: 1, BCH: 2 }
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply constants/index.ts:100]: indeed! done

[the remark this answers is not in the document]
Comment (vorg-code) [reply constants/index.ts:100]: also for QUOTES

File: prisma/seeds/prices.ts [hunk 10]
```diff
@@ -1,35 +1,127 @@
...
+    return 0
+  })
+  await writeToFile(PATH_PRICE_CSV_FILE, prices)
+
+  const finish = moment()
+  console.log(`\n\nstart: ${start.format('HH:mm:ss')}\nfinish: ${finish.format('HH:mm:ss')}\nduration: ${(finish.diff(start) / 1000).toFixed(2)} seconds`)
+}
+
+async function readCsv (filePath: string): Promise<string[][]> {
+  return await new Promise((resolve, reject) => {
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply prisma/seeds/prices.ts:61]: There is, since the `fs.readFile` is not async await (it uses callback), we have to wrap it in a promise to be able to wait on it upon the method `readCsv` call

File: prisma/seeds/prices.ts [hunk 11]
```diff
@@ -1,35 +1,127 @@
...
+          data
+            .split('\n')
+            .map((row) => row.split(','))
+        )
+      }
+    })
+  })
+}
+
+async function fileExists (filePath: string): Promise<boolean> {
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply prisma/seeds/prices.ts:76]: indeed, done

File: prisma/seeds/prices.ts [hunk 12]
```diff
@@ -1,35 +1,127 @@
...
+  priceInCAD: string
+  priceInUSD: string
+}
+
+export const PATH_PRICE_CSV_FILE = path.join('prisma', 'seeds', 'prices.csv')
+
+async function writeToFile (fileName: string, content: PriceFileData[]): Promise<void> {
+  const writeFile = promisify(fs.writeFile)
+
+  const headers = ['ticker', 'date', 'priceInCAD', 'priceInUSD']
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply prisma/seeds/prices.ts:22]: Agreed, that would be a nice solution, but we have conflicts in package versions when trying to install this package: 
photo_5152556271285677094_y

It seems it could be quite a lot of work or version break something to avoid repeating the interface definition here

File: prisma/seeds/prices.ts [hunk 13]
```diff
@@ -1,35 +1,127 @@
...
+  try {
+    await fs.promises.stat(filePath)
+    return true
+  } catch (error) {
+    return false
   }
-]
+}
+
+async function getPricesFromFile (attempt: number = 1): Promise<PriceFileData[]> {
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply prisma/seeds/prices.ts:85]: - if I do the reduce inside the `readCsv`, then I would have to take it out of `utils/index.ts` as it wouldn't be reusable anymore
- indeed, refactored

File: prisma/seeds/prices.ts [hunk 14]
```diff
@@ -1,35 +1,127 @@
...
+    const csvContent = await readCsv(PATH_PRICE_CSV_FILE)
+    const res: PriceFileData[] = []
+
+    for (let i = 1; i < csvContent.length; i++) {
+      const newPrice: KeyValueAny = {}
+      for (let j = 0; j < csvContent[0].length; j++) { // headers
+        newPrice[csvContent[0][j]] = csvContent[i][j]
+      }
+      // ignore some prices that come empty
+      if (newPrice.priceInCAD !== '' && newPrice.priceInUSD !== '') { res.push(newPrice as PriceFileData) }
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply prisma/seeds/prices.ts:96]: indeed there are such few datapoints where this happens, idk why, since I don't have access to their database, here are some examples:
image

File: services/priceService.ts [hunk 15]
```diff
@@ -62,23 +61,121 @@ export async function upsertCurrentPricesForNetworkId (responseData: IResponseDa
...
   }
-  const todayString = moment().format(PRICE_API_DATE_FORMAT)
+  if (appInfo.priceAPIToken === '') {
+    throw new Error(RESPONSE_MESSAGES.MISSING_PRICE_API_TOKEN_400.message)
+  }
+}
+
+function getPriceByDayURL (day: moment.Moment, ticker: string): string {
+  validatePriceAPIUrlAndToken()
+  validateTicker(ticker)
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply services/priceService.ts:85]: Nice, moved them both to validators. 

About the rename, I don't agree with it, because it only validates if the ticker is a ticker we support, i.e. if it is in our const list of tickers. If that ticker has a network or quote related to it or not is a separate issue.

File: services/priceService.ts [hunk 16]
```diff
@@ -62,23 +61,121 @@ export async function upsertCurrentPricesForNetworkId (responseData: IResponseDa
...
+  return `${appInfo.priceAPIURL}/pricebydate/${appInfo.priceAPIToken}/${ticker}+${day.format(PRICE_API_DATE_FORMAT)}`
+}
+
+function getAllPricesByTickerURL (ticker: string): string {
+  validatePriceAPIUrlAndToken()
+  validateTicker(ticker)
+  return `${appInfo.priceAPIURL}/dailyprices/${appInfo.priceAPIToken}/${ticker}`
+}
+
+export async function getPriceForDayTicker (day: moment.Moment, ticker: string, attempt: number = 1): Promise<IResponseData | null> {
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply services/priceService.ts:95]: I used `ticker` here because the api only accepts the `ticker`, as an input and also it outputs the ticker, there is no `id` to be used in these methods. 

image

File: services/priceService.ts [hunk 17]
```diff
@@ -62,23 +61,121 @@ export async function upsertCurrentPricesForNetworkId (responseData: IResponseDa
...
+  try {
+    const res = await axios.get(getPriceByDayURL(day, ticker), {
+      timeout: PRICE_API_TIMEOUT
+    })
+
+    if (res.data.success !== false) { return res.data } else { return null }
+  } catch (error) {
+    console.error(`Problem getting price of ${ticker} ${day.format('YYYYMMDD')} -> ${error as string} (attempt ${attempt})`)
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply services/priceService.ts:103]: done

File: services/priceService.ts [hunk 18]
```diff
@@ -62,23 +61,121 @@ export async function upsertCurrentPricesForNetworkId (responseData: IResponseDa
...
+  const lastTimestamp = await prisma.price.findFirst({
+    where: {
+      networkId: XEC_NETWORK_ID, // any valid network
+      quoteId: USD_QUOTE_ID // any valid quote
+    },
+    orderBy: [{ timestamp: 'desc' }],
+    select: { timestamp: true }
+  })
+
+  if (lastTimestamp?.timestamp === undefined) throw new Error('No prices found in the worker, please run prisma seed')
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply services/priceService.ts:145]: indeed! both adressed

File: services/priceService.ts [hunk 19]
```diff
@@ -62,23 +61,121 @@ export async function upsertCurrentPricesForNetworkId (responseData: IResponseDa
...
+          day: e[0],
+          ...e[1]
+        }
+      })
+      return dailyPrices
+    } else { return null }
+  } catch (error) {
+    console.error(`Problem getting price of ${ticker} -> ${error as string} (attempt ${attempt})`)
+
+    if (attempt < PRICE_API_MAX_RETRIES) {
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply services/priceService.ts:129]: indeed, random subconscious non-blocking mentality I guess, error thrown

File: services/priceService.ts [hunk 20]
```diff
@@ -62,23 +61,121 @@ export async function upsertCurrentPricesForNetworkId (responseData: IResponseDa
...
+      return await getAllPricesByTicker(ticker, attempt + 1)
+    } else { return null }
+  }
+}
+
+export async function syncPastPrices (): Promise<void> {
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply services/priceService.ts:135]: changed it to the more descriptive `syncPastDaysNewerPrices`

File: services/priceService.ts [hunk 21]
```diff
@@ -62,23 +61,121 @@ export async function upsertCurrentPricesForNetworkId (responseData: IResponseDa
...
+  while (date.isAfter(lastDateInDB)) {
+    datesToRetrieve.push(date.clone())
+    date.add(-1, 'day')
+  }
+
+  await Promise.all(
+    Object.values(TICKERS).map(async (ticker) =>
+      datesToRetrieve.map(async date => {
+        const price = await getPriceForDayTicker(date, ticker)
+        if (price != null) {
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply services/priceService.ts:160]: The `getPriceForDayTicker` already tries many times to avoid network error, but if the api gives us a null price because it is broken for that day or doesn't have its data, then throwing an error here would kill the worker and avoid syncing the other days correct prices that come back from the api, that `if` allows that. 

Since we can't automatically do nothing if the api doesn't give us the price, better have 1 day hole than 5 day hole.

That being said, I will include a `console.error` message here to warn about this.

**Comment by vorg-code** [comment]
@ScottMcDermid Can you post here the price file CSV (/prisma/seed/prices.csv) that you have on your machine that triggered this first error? Here it doesn't happen, it seems like an empty price that escaped my empty treatment. 

As for the second error, it is from the worker syncUnsyncedAddressesWorker that was not touched in this PR. Probably you got this error because of the first error that stopped the database from getting the basic data that it needs, so it was empty and this worker failed because this networkId was not found in the database.

File: prisma/seeds/prices.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (chedieck) [reply prisma/seeds/prices.ts:22]: Alright, best that we update the node in another issue and maybe refactor it then

File: prisma/seeds/prices.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (chedieck) [reply prisma/seeds/prices.ts:61]: I believe you could use `fs.promises.readFile` as you did with `fs.promises.stat`

File: services/priceService.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (chedieck) [reply services/priceService.ts:85]: I don't understand, those tickers are network tickers, not quote tickers. E.g: XEC is a network ticker; in the future could be a quote ticker also (meaning e.g. we could show the dashboard value of all paybuttons in XEC). This const list of tickers are the tickers of the networks we support. `validateTicker` is going to fail for `USD`, which is a quote ticker we support.

File: services/priceService.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (chedieck) [reply services/priceService.ts:95]: Alright, makes sense

File: services/priceService.ts (hunk 5 above)
Comment (vorg-code) [reply services/priceService.ts:83]: Done

File: prisma/seeds/prices.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (vorg-code) [reply prisma/seeds/prices.ts:61]: indeed! done

File: services/priceService.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (vorg-code) [reply services/priceService.ts:85]: Now I understood better what you meant, makes sense, done

**Comment by vorg-code** [comment]
All of @ScottMcDermid and @chedieck code remarks addressed.

About the points @chedieck raised outside code lines:

1. I like the trunk based philosophy, I will try my best to adapt to it in the next branches
2. Since this price load has to be solved before the environment goes up, it should not be a job task imo, tasks are kinda independent from the main code, executing whenever, here we don't have this luxury. Also, we never append or fix this file, we create only once and then we never touch it again, this is just to seed the database, further price syncs or holes will be filled in by the job that runs afterwards, then indeed we do the 3 steps you mentioned.
3. Indeed, there are 3 recursive limited number of attempts when: building the file from the api, getting a price from the api for a day and getting all daily prices for a ticker from the api. First item calls the last one, so extra resilience there. My fear here was a network time out or corrupted data that can happen for any one at any time causing the request to crash with an exception or the whole environment not to be started, when we could simply retry and it would work seamlessly for the user. I also had to limit the attempts, since we can't retry forever and let the system hang if the api is down for example. If you know a better way to accomplish these goals, I am all in for the learning 😃 
4. I didn't get this one... when syncing prices we are accessing an external api also right?

**Comment by ScottMcDermid** [comment]
> @ScottMcDermid Can you post here the price file CSV (/prisma/seed/prices.csv) that you have on your machine that triggered this first error? Here it doesn't happen, it seems like an empty price that escaped my empty treatment.
> 
> As for the second error, it is from the worker syncUnsyncedAddressesWorker that was not touched in this PR. Probably you got this error because of the first error that stopped the database from getting the basic data that it needs, so it was empty and this worker failed because this networkId was not found in the database.

Yes, here's my prices.csv after deleting the 3 last XEC lines:
[prices.csv](https://github.com/PayButton/paybutton-server/files/10796423/prices.csv)

**Comment by vorg-code** [comment]
@chedieck:
1. Method name changed to `syncCurrentAndPastPricesWorker` (logging text adjusted as well)
2. Since we now are using `getAllPricesByTicker`, I think that either we will get all prices or none, so searching for missing days and retrying doesn't make much sense to me. This method is one of the 3 retries one, so we should eventually get the full list of prices. There are only 3 functions that retry, these 2 (`getPriceForDayTicker` and `getAllPricesByTicker`) because they call the external api and I guess the last one (`getPricesFromFile`) was unnecessary to do it like this. I removed the multiple attempts there.
3. I changed the function's name to `getPriceURLForDayAndNetworkTicker`, however, it does not seem like a good idea to me to use the name `syncPriceForDayAndNetworkTicker`, since this function is not syncing anything, it just gets the price from an external source and returns. In contrast, for example, `syncCurrentPrices` indeed syncs what we have in the database with what we have in the external source. If you feel the need to indicate in the function's name `getPriceForDayTicker` that it accesses some external source, maybe we can call it `getExternalPriceForDayAndNetworkTicker` ?

@ScottMcDermid what happened in your file is that you left an empty row in its last line:
image
I altered the code to account for empty lines without issues.

**Review by chedieck — CHANGES_REQUESTED** [review]

File: prisma/seeds/prices.ts [hunk 22]
```diff
@@ -1,35 +1,105 @@
...
+async function getPricesFromFile (attempt: number = 1): Promise<PriceFileData[]> {
+  if (await fileExists(PATH_PRICE_CSV_FILE)) {
+    const csvContent = await readCsv(PATH_PRICE_CSV_FILE)
+    const res: PriceFileData[] = []
+
+    const headers = csvContent[0]
+    const data = csvContent.slice(1)
+
+    for (const line of data) {
+      const newPrice: KeyValueT<any> = {}
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply prisma/seeds/prices.ts:69]: Changed it to `KeyValueT<string>`, the one you suggested can't be set because all attributes are mandatory and I am initializing it as an empty object here.

About the outdated conversation below about throwing an error if one of the quotes come back missing from the api, I think we should do that when the data comes back from the api and not when reading the file (user can manually change the file and we don't wanna create the file with these wholes then). This was addressed there.

File: constants/index.ts [hunk 23]
```diff
@@ -71,11 +77,24 @@ export const SUPPORTED_QUOTES = [ // avoids hitting the DB every time for data t
...
+export const PRICE_FILE_MAX_RETRIES = 3
 
 export const BCH_TIMESTAMP_THRESHOLD = 1501588800 // 2017 Aug 1, 12PM
 export const XEC_TIMESTAMP_THRESHOLD = 1605398400 // 2020 Nov 15, 12AM
 
 // Wait time (in ms) between sync of current prices
 export const CURRENT_PRICE_SYNC_DELAY = 60000
 export const DEFAULT_WORKER_LOCK_DURATION = 120000
+
+export const TICKERS: KeyValueT<string> = {
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply constants/index.ts:93]: done

File: services/priceService.ts [hunk 24]
```diff
@@ -62,23 +62,106 @@ export async function upsertCurrentPricesForNetworkId (responseData: IResponseDa
...
+  await upsertPricesForNetworkId(responseData, networkId, 0)
+}
+
+function getURLForPriceByDayRequest (day: moment.Moment, ticker: string): string {
+  validatePriceAPIUrlAndToken()
+  validateNetworkTicker(ticker)
+  return `${appInfo.priceAPIURL}/pricebydate/${appInfo.priceAPIToken}/${ticker}+${day.format(PRICE_API_DATE_FORMAT)}`
+}
+
+function getAllPricesByTickerURL (ticker: string): string {
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply services/priceService.ts:75]: changed it to `getAllPricesURLForTicker`

File: services/priceService.ts [hunk 25]
```diff
@@ -62,23 +62,106 @@ export async function upsertCurrentPricesForNetworkId (responseData: IResponseDa
...
+  validatePriceAPIUrlAndToken()
+  validateNetworkTicker(ticker)
+  return `${appInfo.priceAPIURL}/dailyprices/${appInfo.priceAPIToken}/${ticker}`
+}
+
+export async function getPriceForDayTicker (day: moment.Moment, ticker: string, attempt: number = 1): Promise<IResponseData | null> {
```
[the remark this answers is not in the document]
Comment (vorg-code) [reply services/priceService.ts:81]: Both addressed.

**Comment by ScottMcDermid** [comment]
> @ScottMcDermid what happened in your file is that you left an empty row in its last line:
> image
> I altered the code to account for empty lines.

I tried rerunning tests, and the results were similar:

Test 1: Populates prices file as expected
Test 2: Runs, and doesn't complain

Now, Test 3 runs without error, but it still doesn't populate the missing prices for me.

Has anyone else experienced this result? I'm wondering if it's an issue exclusive to my environment.

**Comment by vorg-code** [comment]
@ScottMcDermid did you deleted the days for both XEC and BCH in the file to do the 3rd step? 

After one of the remarks, I removed the filter for the network_ticker when fetching the last day that we have prices for in the database, so if you just remove the last days from XEC at the end of the file but leave the same removed days in BCH, then indeed nothing will happen in the 3rd step.

**Comment by ScottMcDermid** [comment]
> @ScottMcDermid did you deleted the days for both XEC and BCH in the file to do the 3rd step?

No, I just deleted the final 3 lines (3 days of XEC prices)

Ok I think this is my mistake. I misinterpreted the expected results for step 3. I believe what I'm seeing matches what you describe in your test plan. Sorry about that.

**Review by ScottMcDermid — APPROVED** [review]
